### PR TITLE
WebDAV backend (WIP)

### PR DESCRIPTION
**Pull request by jlbprof**

### CODE REVIEW

**Comment by jlbprof** [comment]
Most of the changes are whitespace, I changed indents to 2 spaces.

I fixed the case where if it fails the lock would not be deleted.

I make sure that the db contains only bookmark objects.

[Image of my developer tools console](https://imgur.com/a/0QNO4PP)

In the test situation where this fails, I have a "test" folder on the bookmarks bar with one bookmark to duckduckgo.   The image shows what my db contains.   That is what would be returned from getBookmark.    I do not know how to proceed to fix this.  I have the "remoteId" set, but the "localId" is not set and do not know what to set that too.   But in essence the remoteId and localId should be the same, and I even tried that but I keep getting the same error about the 2nd bookmark with the same url.  And it is trying to create that 2nd copy of the url.    I am going to need help debugging this, because this is probably a dance between your side and my side that is not going well.

When we get this fixed, I will work the "path" problem with slashes.

Thanx

Julian

File: src/lib/adapters/WebDav.js [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+    bm.title,
+    bm.path));
+
+  console.log('Create bookmark: OUT ', bm, this.server);
+
+  return bm;
+}
+
+async updateBookmark (remoteId, newBm) {
+  console.log('Update bookmark', newBM, remoteId, this.server)
```
Comment (marcelklehr) [inline src/lib/adapters/WebDav.js]: This throws a `newBM is not defined` Error :)

**Comment by marcelklehr** [comment]
Ah, I see the problem now. So, when floccus is syncing from scratch without any mappings available yet. It tries to create all bookmarks that are in the local sync folder on the server as well. The bookmarks app has the constraint that every URL may only appear once, so instead of creating a new bookmark it modifies the existing one and returns the existing ID. Floccus relies on this to initialize its mappings.

There are different ways forward from this. We can try to move the code that deals with the peculiarities of the bookmarks app into the Adapter and have floccus work on trees with folder ids and duplicate bookmarks instead of the flat hierarchy of the bookmarks app. Or we can mimick the bookmarks app with the WebDAV adapter and also prevent duplicates.

**Comment by TeutonJon78** [comment]
Seems a more robust way would be keep any backend specific stuff in an Adapter. That way, if you add more backends in the future, they has the solid core of floccus to build off, rather than having to adapt to the Bookmarks app conventions.

**Comment by jlbprof** [comment]
My recommendation is that we redesign the adapter to be a "tree" style
adapter.  In the case of Nextcloud we would "flatten" the tree, but for any
adapter that can represent a tree like WebDav and XBel we can "enrich" the
user experience.   Some of the biggest issues I have with Floccus is that I
cannot sort the order of the bookmarks.   I think it would be an
appreciated feature.   Nextcloud cannot do the tree, I understand that, so
it would have to deal with not allowing duplicate url's or keep the order
of bookmarks across browsers.

Let me know how to proceed on that and I will do so.   If we do allow for a
new tree structure we should probably discuss the adapter changes, such as
methods and parameters.

I will fix the other parts about, path and the strings as an integer
problem tomorrow.

Julian

**Comment by marcelklehr** [comment]
@TeutonJon78 @jlbprof 
> My recommendation is that we redesign the adapter to be a "tree" style
adapter. 

I agree that's probably the more sustainable option. 

@jlbprof I've been working on a UI refactoring which I would like to merge before we refactor the Adapter interface in general, so we spare ourselves some rebasing :D So before that, I'd like to merge this PR as-is once you've fixed the last bits other than the problem I described; then I'll merge my UI refactoring, adjust your adapter code to work with the new UI framework; then we work out the details on how to revamp the adapter interface. Perhaps, while I do the UI stuff, you can already sketch an interface that would suit your adapter as a first draft, on a wiki page or in an issue.

**Comment by jlbprof** [comment]
I hope to have the changes in by tomorrow evening.

**Comment by jlbprof** [comment]
Thanx

**Comment by jlbprof** [comment]
Been thinking about the redesign with a tree style adapter.

I think the adapter would be interacted with via 3 methods.

syncStart
syncFail
syncComplete

When you call syncStart it will return and entire Tree object with the
server's bookmarks.
When you call syncComplete you pass it a Tree object, that would include
any changes you have made such as:
* Removed folders or bookmarks would be flagged
* Changes to title, url would be flagged.

That way the Nextcloud adapter would just pass through "the changed"
sections and make appropriate changes on the server.
The other adapters would output a new bookmark file with the changes in
place.

This would make the adapter trivial, your job as the sync process would be
more complex perhaps.

Because of the way I built the WebDav, I could have some more options

1) Output in XBEL format.
2) Output in bookmarks html formt.

Just something to think about.

Julian

**Comment by marcelklehr** [comment]
The syncComplete wouldn't necessarily need to be a tree, though, right? It could just be an array of operations with nodes referenced by id, no?

**Comment by jlbprof** [comment]
A tree is just a suggestion.   If the list of nodes show a hierarchy that
would be fine too.

**Pull request merged**